Update XCFramework pods after running `bin/generate-podspecs.sh`

mokagio opened the pull request:
@jhnstn recently opened https://github.com/wordpress-mobile/gutenberg-mobile/pull/5843 to address a `bundle exec pod install` failing locally for the XCFramework generation project.

As far as I can tell, the sequence of events that resulted in that failure was:

1. Some RN dependencies were updated in Gutenberg via https://github.com/WordPress/gutenberg/pull/50731
2. Those new versions were eventually fetched in this repo (I'm guessing via `bin/generate-podspec.js`?) in https://github.com/wordpress-mobile/gutenberg-mobile/commit/ed9625d394b117fc9271fdf642ff9167e95d03cd
3. Suddenly, the version in the XCFramework `Podfile.lock` was pointing to one that no longer existed locally

Assuming that `bin/generate-podspec.js` is the single point through which we updated the `third-party-podspecs`, then I think we can avoid this issue in the future by making the script also update the pods used by the XCFramework project, which is what this PR does.

### Testing

I tried to run the script locally, but it failed like so:

```
WARNING! hermes-engine.podspec doesn't have a 'tag' or 'commit' field, or doesn't point to a SHA verified file. Either modify this script to add a patch during the podspec generation or modify the original hermes-engine.podspec in the source repo.
```

<details>
<summary>Full log</summary>

```
➜ ./bin/generate-podspecs.sh
Working directory: /Users/gio/Developer/a8c/gutenberg-mobile
If your node_modules folder isn't up-to-date please run 'npm install' first and re-run the script. Is your node_modules folder up-to-date? [y/N] y
Enter the commit hash of previous commit. If this is the first-time running this script, enter 0, then commit generated files and re-rerun the script and this time use the previous commit hash: ddd8905b097ace9177cb2aa5d5414c923ccb482a
Please delete '/Users/gio/Developer/a8c/gutenberg-mobile/third-party-podspecs' folder manually before continuing. This script will re-generate it. Did you delete it? [y/N] y
Generating podspec for boost
Generating podspec for RCT-Folly
Generating podspec for DoubleConversion
Generating podspec for glog
Generating podspec for react-native-blur
   ==> Patching react-native-blur podspec
Generating podspec for RNCMaskedView
Generating podspec for react-native-slider
Generating podspec for RNCClipboard
Generating podspec for RNGestureHandler
Generating podspec for react-native-get-random-values
Generating podspec for BVLinearGradient
Generating podspec for RNReanimated
Generating podspec for react-native-safe-area
Generating podspec for react-native-safe-area-context
Generating podspec for RNScreens
Generating podspec for RNSVG
Generating podspec for react-native-video
Generating podspec for react-native-webview
Generating podspec for RNFastImage
Generating podspec for React-RCTVibration with path Libraries/Vibration
Generating podspec for React-RCTSettings with path Libraries/Settings
Generating podspec for React-RCTAnimation with path Libraries/NativeAnimation
Generating podspec for FBLazyVector with path Libraries/FBLazyVector
Generating podspec for React-RCTActionSheet with path Libraries/ActionSheetIOS
Generating podspec for React-RCTNetwork with path Libraries/Network
Generating podspec for RCTRequired with path Libraries/RCTRequired
Generating podspec for React-RCTImage with path Libraries/Image
Generating podspec for RCTTypeSafety with path Libraries/TypeSafety
Generating podspec for React-RCTLinking with path Libraries/LinkingIOS
Generating podspec for React-RCTText with path Libraries/Text
Generating podspec for React-RCTBlob with path Libraries/Blob
Generating podspec for React-RCTPushNotification with path Libraries/PushNotificationIOS
Generating podspec for React-CoreModules with path React/CoreModules
Generating podspec for FBReactNativeSpec with path React/FBReactNativeSpec
/Users/gio/Developer/a8c/gutenberg-mobile/gutenberg/node_modules/react-native/scripts/react_native_pods_utils/script_phases.rb:51: warning: Passing safe_level with the 2nd argument of ERB.new is deprecated. Do not use it, and specify other arguments as keyword arguments.
/Users/gio/Developer/a8c/gutenberg-mobile/gutenberg/node_modules/react-native/scripts/react_native_pods_utils/script_phases.rb:51: warning: Passing trim_mode with the 3rd argument of ERB.new is deprecated. Use keyword argument like ERB.new(str, trim_mode: ...) instead.
   ==> Patching FBReactNativeSpec podspec
Generating React-Codegen
Generating schema from Flow types
Generating native code from schema (iOS)
Generating podspec for React-Core with path .
Generating podspec for React with path .
Generating podspec for React-jsiexecutor with path ReactCommon/jsiexecutor
Generating podspec for Yoga with path ReactCommon/yoga
Generating podspec for React-logger with path ReactCommon/logger
Generating podspec for React-jsinspector with path ReactCommon/jsinspector
Generating podspec for React-bridging with path ReactCommon
Generating podspec for React-cxxreact with path ReactCommon/cxxreact
Generating podspec for React-jsi with path ReactCommon/jsi
Generating podspec for React-perflogger with path ReactCommon/reactperflogger
Generating podspec for React-runtimeexecutor with path ReactCommon/runtimeexecutor
Generating podspec for ReactCommon with path ReactCommon
Generating podspec for React-hermes with path ReactCommon/hermes
Generating podspec for React-callinvoker with path ReactCommon/callinvoker
Generating podspec for React-graphics with path ReactCommon/react/renderer/graphics
Generating podspec for hermes-engine with path sdks/hermes-engine
WARNING! hermes-engine.podspec doesn't have a 'tag' or 'commit' field, or doesn't point to a SHA verified file. Either modify this script to add a patch during the podspec generation or modify the original hermes-engine.podspec in the source repo.
```

</details>

I'm not sure whether the failure is due to the SHA1 I used or there's something currently wrong with `hermes-engine`.

---


PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/RELEASE-NOTES.txt) if necessary.

mokagio committed:
Wrap a shell script var in `""` to be robust against spaces

mokagio committed:
Update XCFramework pods after running `bin/generate-podspecs.sh`

The XCFramework pods refer to the local third-party podspecs. When those
are updated, it can happen that `bundle exec pod install` cannot find
the new version. Running `bundle exec pod update` ensures that all
dependencies point to the version defined in the local third-party
podspecs.

dcalhoun committed:
build: Fix script directory changes

Rely upon pushd and popd to simplify the "breadcrumbs" of navigating
into and out of multiple directories.

dcalhoun commented on bin/generate-podspecs.sh:
```diff
@@ -195,4 +195,9 @@ do
+echo 'Updating XCFramework Podfile.lock with these changes'
+pushd ios-xcframework
+bundle exec pod update
+popd
```
I ran into a few errors when testing this addition. 

```
./bin/generate-podspecs.sh: line 201: pushd: ios-xcframework: No such file or directory
```

The above directory error is from the directory change failing because previous changes of directory were not reset.  

```
Installing FBReactNativeSpec 0.69.4

[!] Error installing FBReactNativeSpec
[!] /opt/homebrew/bin/git -C /var/folders/c5/mkbb257d7fg4myx88dvsndp80000gn/T/d20230623-6856-98pfen checkout --quiet 0

error: pathspec '0' did not match any file(s) known to git
```

The above `pathspec` errors results from my understanding is that `bin/generate-podspecs.sh` has to be run twice for each change (p9ugOq-29e-p2). You can observe the result of the second run for https://github.com/wordpress-mobile/gutenberg-mobile/pull/5789 in https://github.com/wordpress-mobile/gutenberg-mobile/commit/bafd5569192641312f9ded9b04028a8ecbcfafbf. You might note the value is changing from `0`, which is a temporary value. That temporary value causes the following error when running the approach outlined in this PR. 

I opened https://github.com/wordpress-mobile/gutenberg-mobile/pull/5901 to resolve these issues.

dcalhoun committed:
build: Only update xcframework Podfile.lock with valid targets

We are required to run `bin/generate-podspecs.sh` twice to capture the
correct target. This change ensures the xcframework is only updated on
the second run with the correct target, otherwise the following error
occurs:

```
Installing FBReactNativeSpec 0.69.4

[!] Error installing FBReactNativeSpec
[!] /opt/homebrew/bin/git -C /var/folders/c5/mkbb257d7fg4myx88dvsndp80000gn/T/d20230623-6856-98pfen checkout --quiet 0

error: pathspec '0' did not match any file(s) known to git
```

dcalhoun reviewed (commented):
Thank you for taking the time to address this issue! 🙇🏻 

> I tried to run the script locally, but it failed like so:
> 
> ```
> WARNING! hermes-engine.podspec doesn't have a 'tag' or 'commit' field, or doesn't point to a SHA verified file. Either modify this script to add a patch during the podspec generation or modify the original hermes-engine.podspec in the source repo.
> ```
> 
> [...]
>
> I'm not sure whether the failure is due to the SHA1 I used or there's something currently wrong with `hermes-engine`.

I circumvented this error by manually setting a `tag` within my local `node_modules` directory. That is obviously not a long-term solution, but it let me proceed. 

I do not believe the `hermes-engine` error relates to this PR specifically, but I could be wrong. For `hermes-engine`, I am unsure whether we need to [patch](https://github.com/wordpress-mobile/gutenberg-mobile/blob/43df7ab8953d5190529c01db7388f5a0683db3e1/bin/generate-podspecs.sh#L78-L86) it, [filter it out](https://github.com/wordpress-mobile/gutenberg-mobile/blob/43df7ab8953d5190529c01db7388f5a0683db3e1/bin/generate-podspecs.sh#L108), or something else. My limited understanding of the purpose[^1] of `bin/generate-podspecs.sh` inhibits me from knowing the best path forward.

[^1]: If you are able to share a high level description of its purpose, that would be very helpful for me. "Explain it to me like I'm five." 😄

dcalhoun committed:
build: Document generate-podspecs script

Co-authored-by: Gio Lodi <gio.lodi@automattic.com>

dcalhoun committed:
build: Improve inline comment grammar and clarity

dcalhoun committed:
Merge pull request #5901 from wordpress-mobile/build/repair-xcframework-integration

build: Fix generate-podspecs script errors

dcalhoun commented:
> 1. If you are able to share a high level description of [the script's] purpose, that would be very helpful for me. "Explain it to me like I'm five." 😄

For posterity, I am answering my own question to the best of my understanding. We have the `./bin/generate-podspecs.sh` script to allow the `WordPress-iOS` project to avoid including any React Native or JavaScript references in its source code or CI tasks. The `WordPress-iOS` project instead points its Podfile to the Podspec files generated in the `gutenberg-mobile` repository, which "manages" all things related to React Native. (ref: p9ugOq-29e-p2)

dcalhoun reviewed (approved):
As mentioned in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5846#pullrequestreview-1495291213, I do not believe the lingering `hermes-engine` error relates to the proposed changes. The error occurs without the proposed changes. Therefore, I believe it is safe to merge this work. 

Thank you for addressing this issue. 🙇🏻
